### PR TITLE
Benchmark Runner for C Benchmarks

### PR DESCRIPTION
**Pull request by mattchorlian**
Opening this draft PR to track progress of converting all C benchmarks to use the benchmark runner:

- [x] Port current C++ benchmark runner to C
- [x] Big.lf
- [x] Chameneos.lf
- [x] Counting.lf
- [x] PingPong.lf
- [x] ThreadRing.lf
- [x] Throughput.lf
- [x] BoundedBuffer.lf
- [x] CigaretteSmoker.lf
- [x] Dictionary.lf
- [x] LogisticMap.lf
- [x] Philosophers.lf
- [x] SleepingBarber.lf
- [x] SortedLinkList.lf
- [x] Apsp.lf
- [x] FilterBank.lf
- [x] GuidedSearch.lf
- [x] MatMul.lf
- [x] NQueens.lf
- [x] PiPrecision.lf
- [x] RadixSort.lf
- [x] Trapezoidal.lf

### CODE REVIEW

**Comment by mattchorlian** [comment]
I'm noticing that we have BoundedBuffer.lf and ProdCons.lf both here even though they are the same benchmark.  I suppose this is (in part) the Savina naming convention causing confusion again.  In any case, we should decide which one to keep around before merging this in!

**Review by cmnrd — CHANGES_REQUESTED** [review]
See the comments I made in #852

**Comment by lhstrh** [comment]
Does this PR also address https://github.com/lf-lang/lingua-franca/issues/765? If not, I think it probably should.

File: benchmark/C/Savina/src/parallelism/GuidedSearch.lf
Comment (lhstrh) [inline benchmark/C/Savina/src/parallelism/GuidedSearch.lf:88]: This file, along with the included files, should be in a separate directory, as is done with the C++ benchmarks.

**Review by lhstrh — CHANGES_REQUESTED** [review]
This looks great, @mattchorlian and @LDeng0205! I've only requested minor changes. Let's merge this into master ASAP.

**Comment by housengw** [comment]
Seems like running `MatMul.lf` with more than 1 thread produces a random validation error. I'm guessing it's a memory issue..

File: benchmark/runner/conf/benchmark/savina_concurrency_concsll.yaml
Comment (cmnrd) [inline benchmark/runner/conf/benchmark/savina_concurrency_concsll.yaml:46]:
```suggestion
    binary: "SortedList"
```

File: .github/workflows/ci.yml
Comment (cmnrd) [inline .github/workflows/ci.yml:29]: This should be reverted to master. There should be no need to change this unless the `benchmark-tests.yaml` workflow itself is updated in this branch.

**Review by cmnrd — COMMENTED** [review]
This looks great! Looks like we are close to completing the C benchmarks.

Should we also include the Banking Benchmark #624 here and fix it up?

> Seems like running MatMul.lf with more than 1 thread produces a random validation error. I'm guessing it's a memory issue.

Could you describe what the issue seems to be? Is it a segfault that occurs sometimes? Or is the result invalid (since you mention a validation error)? The MatMul benchmark has a built-in race condition and hence the result is expected to be invalid for parallel execution. This is a bug in the original Savina benchmark...

**Comment by housengw** [comment]
the `validation error` in multithreaded run of `MatMul.lf` is likely to be a race condition.

**Comment by lhstrh** [comment]
> the `validation error` in multithreaded run of `MatMul.lf` is likely a race condition.

Could you elaborate?

**Comment by housengw** [comment]
The MatMul benchmark has random validation errors when running with more than 1 thread. Since there is no segfault and the validation error happens at random indices in the matrix, it is most likely a race condition. 
As @cmnrd mentioned, the MatMul benchmark has a built-in race condition, so running it with multiple threads would cause such an error.

**Review by lhstrh — APPROVED** [review]
Looks ready to me! This is a big feat and will be very useful for the evaluation of the new schedulers in `reactor-c` 🚀🚀🚀 Thanks!

**Comment by mattchorlian** [comment]
> Should we also include the Banking Benchmark #624 here and fix it up?

We decided to get this merged first and deal with Banking separately @cmnrd

**Comment by lhstrh** [comment]
> > Should we also include the Banking Benchmark #624 here and fix it up?
> 
> We decided to get this merged first and deal with Banking separately @cmnrd

Comt to thing of it, there is one distinct advantage to adding the other stuff to this PR: we'll have a cleaner history and release notes (if we automatically generate them). Doesn't seem like it would be much work?

**Review by cmnrd — APPROVED** [review]
Looks great! I am happy with both options, merging as is or including banking as well.

**Comment by LDeng0205** [comment]
I just added benchmark runner to Banking as well so we could include it with this PR. However, I still have a   java.lang.OutOfMemoryError when I compile it with NumAccounts = 1000.

**Comment by edwardalee** [comment]
Is this branch merged with master? It looks like it is still generating in-line code, which suggests it is using a version of master that predates scalability-banks.

**Pull request merged**
